File: 0238-product-of-array-except-self/solution.java
class Solution {
   public static int[] productExceptSelf(int[] nums) {
      int totalProduct = 1;
      boolean noZeros = true;
      boolean allZeros = false;
      int zeroCount = 0;
      int[] res = new int[nums.length];
      for (int i = 0; i < nums.length; i++) {
         if (nums[i] != 0) {
            totalProduct *= nums[i];
            continue;
         }
         zeroCount++;
         noZeros = false;
      }
      if (zeroCount > 1) {
         return res;
      }
      
      for (int j = 0; j < nums.length; j++) {
         if (noZeros == false && nums[j] != 0) {
            res[j] = 0;
            continue;
         }
         if (noZeros == false && nums[j] == 0) {
            res[j] = totalProduct;
            continue;
         }
         res[j] = totalProduct / nums[j];
      }
      return res;
   }
    
   public static void main(String[] args) {
      int[] test = {0,4,0};
      int[] res = Solution.productExceptSelf(test);
      for (int i = 0; i < res.length; i++) {
         System.out.println(res[i]);
      }
   }
}

File: 0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        currProduct = 1
        prefix = [1] * len(nums)
        suffix = [1] * len(nums)
        
        for i in range(1, len(prefix)):
            prefix[i] = prefix[i-1] * nums[i - 1]
            

        for j in range(len(prefix)-2, -1, -1):
            suffix[j] = suffix[j+1] * nums[j+1]
        
        print(prefix)
        print(suffix)
            
        for x in range(len(nums)):
            prefix[x] *= suffix[x]
            print(prefix[x])
        
        return prefix
